forgerock-auth/src/main/java/org/forgerock/android/auth/WebAuthnDataRepository.kt

```diff
@@ -0,0 +1,151 @@
+import org.forgerock.android.auth.webauthn.PublicKeyCredentialSource.Companion.fromJson
+import org.json.JSONArray
+import org.json.JSONObject
+
+/**
+ * A Repository that store [PublicKeyCredentialSource]
+ */
+
+private val TAG = WebAuthnDataRepository::class.java.simpleName
+internal const val ALLOW_CREDENTIALS = "ALLOW_CREDENTIALS"
+private const val ORG_FORGEROCK_V_1_WEBAUTHN_KEYS = "org.forgerock.v1.WEBAUTHN_KEYS"
+private const val ORG_FORGEROCK_V_1_WEBAUTHN = "org.forgerock.v1.WEBAUTHN"
+private const val ORG_FORGEROCK_V_2_WEBAUTHN = "org.forgerock.v2.WEBAUTHN"
+
+open class WebAuthnDataRepository internal constructor(val context: Context,
+                                                       sharedPreferences: SharedPreferences?) {
+    private lateinit var dataRepository: SharedPreferences
+
+    init {
+        try {
+            dataRepository = sharedPreferences
+                ?: SecuredSharedPreferences(context,
+                    ORG_FORGEROCK_V_1_WEBAUTHN,
+                    ORG_FORGEROCK_V_1_WEBAUTHN_KEYS)
+            dataRepository = this.migrate(dataRepository)
+        } catch (error: Error) {
+            //Something wrong cannot support Usernameless
+            error(TAG, error, "UsernameLess cannot be supported.")
+        }
+    }
+
+    internal open fun getNewSharedPreferences(): SharedPreferences {
+        return EncryptedPreferences.getInstance(context, ORG_FORGEROCK_V_2_WEBAUTHN)
+    }
+
+    private fun migrate(sharedPreferences: SharedPreferences): SharedPreferences {
+        val encryptedPreferences = getNewSharedPreferences()
+        sharedPreferences.getString(ALLOW_CREDENTIALS, null)?.apply {
+            val map = mutableMapOf<String, MutableSet<String>>()
+            val array = JSONArray(this)
+            for (o in array.iterator<JSONObject>()) {
+                fromJson(o).apply {
+                    (map[this.rpid] ?: mutableSetOf<String>().also {
+                        map[this.rpid] = it
+                    }).add(this.toJson().put("created", System.currentTimeMillis()).toString())
+                }
+            }
+            encryptedPreferences.edit().apply {
+                map.forEach {
+                    this.putStringSet(it.key, it.value)
+                }
+            }.apply()
+
+            sharedPreferences.edit().remove(ALLOW_CREDENTIALS).apply()
+        }
+        return encryptedPreferences
+    }
+
+    /**
+     * Persist the [PublicKeyCredentialSource], it overrides the existing
+     * [PublicKeyCredentialSource] with the same [PublicKeyCredentialSource.getUserHandle]
+     * * @param source The [PublicKeyCredentialSource] to persist.
+     */
+    fun persist(source: PublicKeyCredentialSource) {
+        if (!this::dataRepository.isInitialized) {
+            warn(TAG, "UsernameLess cannot be supported. No credential will be stored")
+            return
+        }
+        dataRepository.getStringSet(source.rpid, emptySet())!!.apply {
```

Comment:
same getStringSet can return null, modify to 

   dataRepository.getStringSet(source.rpid, emptySet())?.apply {
}

Reply:
I have added the null check, but the null check is not really necessary, cause the default is emptySet instead of null.